src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs

```diff
@@ -0,0 +1,295 @@
+        }
+
+        /// <summary>
+        /// Asynchronously ends a scope for writing a stream value.
+        /// </summary>
+        /// <returns>A task representing the asynchronous operation.</returns>
+        public async Task EndStreamValueScopeAsync()
+        {
+            await this.binaryValueStream.FlushAsync();
+            await this.binaryValueStream.DisposeAsync();
+            this.binaryValueStream = null;
+            await this.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+
+            this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+        }
+
+        /// <summary>
+        /// Represents a stream for writing UTF-8 JSON data to an ODataUtf8JsonWriter.
+        /// </summary>
+        internal class ODataUtf8JsonWriterStream : Stream
+        {
+            ODataUtf8JsonWriter jsonWriter = null;
+            private byte[] buffer;
+
+            /// <summary>
+            /// Initializes a new instance of the ODataUtf8JsonWriterStream class with the specified ODataUtf8JsonWriter.
+            /// </summary>
+            /// <param name="writer">The OData UTF-8 JSON writer to write to.</param>
+            internal ODataUtf8JsonWriterStream(ODataUtf8JsonWriter writer)
+            {
+                this.jsonWriter = writer;
+            }
+            public override bool CanRead => false;
+
+            public override bool CanSeek => false;
+
+            public override bool CanWrite => true;
+
+            public override long Length => throw new NotImplementedException();
+
+            public override long Position { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
+
+            /// <summary>
+            /// Flushes any buffered data to the underlying stream synchronously.
+            /// </summary>
+            public override void Flush()
+            {
+                // If there are unprocessed bytes, encode and write them as the final block.
+                ReadOnlySpan<byte> bytesNotProcessedFromPreviousChunk = buffer;
+
+                if (!bytesNotProcessedFromPreviousChunk.IsEmpty)
+                {
+                    this.jsonWriter.Base64EncodeAndWriteChunk(bytesNotProcessedFromPreviousChunk, isFinalBlock: true, out int bytesNotWrittenFromPreviousChunk);
+                    // Clear the buffer since all bytes have been written.
+                    buffer = Array.Empty<byte>();
+                }
+
+                this.jsonWriter.Flush();
+            }
+
+            /// <summary>
+            /// Flushes any buffered data to the underlying stream asynchronously.
+            /// </summary>
+            /// <param name="cancellationToken">A CancellationToken to observe while waiting for the flush operation to complete.</param>
+            /// <returns>A task representing the asynchronous flush operation.</returns>
+            public override async Task FlushAsync(CancellationToken cancellationToken)

```

Comment:
Are we really observing the cancellation token while waiting for the flush operation to complete?

Reply:
Maybe there would be an opportunity to handle the cancellation if `IJsonWriter.FlushAsync` accepted a cancellation token.